test/e2e/operator_test.go

```diff
@@ -228,3 +233,419 @@ func setImage(deployment *appsv1.Deployment, image string) {
+	}
+	defaultDNSPods := &corev1.PodList{}
+	if err := cl.List(context.TODO(), defaultDNSPods, client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		t.Fatalf("failed to list pods for dns daemonset %s/%s: %v", dnsDaemonSet.Namespace, dnsDaemonSet.Name, err)
+	}
+	catCmd := []string{"cat", "/etc/coredns/Corefile"}
+	dnsContainer := "dns"
+	for _, pod := range defaultDNSPods.Items {
+		if err := lookForStringInPodExec(pod.Namespace, pod.Name, dnsContainer, catCmd, upstreamIP, 2*time.Minute); err != nil {
+			t.Fatalf("failed to lookup Corefile in pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		}
+	}
+
+	// Get the openshift-cli image.
+	dnsOp := &configv1.ClusterOperator{}
+	opName := "dns"
+	if err := cl.Get(context.TODO(), types.NamespacedName{Name: opName}, dnsOp); err != nil {
+		t.Fatalf("failed to get clusteroperator %s: %v", opName, err)
+	}
+	var (
+		image string
+		found bool
+	)
+	if dnsOp.Status.Versions != nil {
+		for _, ver := range dnsOp.Status.Versions {
+			if ver.Name == "openshift-cli" {
+				image = ver.Version
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		t.Fatalf("failed to find openshift-cli version in clusteroperator %s: %v", opName, err)
+	}
+
+	// Create the client Pod.
+	testClient := clientPod("test-client", "default", image, []string{"sleep", "3600"})
+	if err := cl.Create(context.TODO(), testClient); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", testClient.Namespace, testClient.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), testClient); err != nil {
+			t.Fatalf("failed to delete pod %s/%s: %v", testClient.Namespace, testClient.Name, err)
+		}
+	}()
+	// Wait for the client Pod to be ready.
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
+		if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: testClient.Namespace, Name: testClient.Name}, testClient); err != nil {
+			return false, nil
+		}
+		for _, cond := range testClient.Status.Conditions {
+			if cond.Type == corev1.ContainersReady &&
+				cond.Status == corev1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to observe ContainersReady condition for pod %s/%s: %v", testClient.Namespace, testClient.Name, err)
+	}
+	// Dig the example dns forwarding host.
+	digCmd := []string{"dig", "+short", "www.foo.com", "A"}
+	fooHost := "1.2.3.4"
+	if err := lookForStringInPodExec(testClient.Namespace, testClient.Name, testClient.Name, digCmd, fooHost, 60*time.Second); err != nil {
+		t.Fatalf("failed to dig %s: %v", upstreamIP, err)
+	}
+	// Scrape the upstream resolver logs for the "NOERROR" message.
+	logMsg := "NOERROR"
+	if err := lookForStringInPodLog(upstreamResolver.Namespace, upstreamResolver.Name, upstreamResolver.Name, logMsg, 60*time.Second); err != nil {
+		t.Fatalf("failed to parse %q from pod %s/%s logs: %v", logMsg, upstreamResolver.Namespace, upstreamResolver.Name, err)
+	}
+}
+
```

Comment:
Consider moving the test to a new file—for example, `test/e2e/forwarding_test.go`—or moving just the utility functions to a new file—for example, `test/e2e/utils.go`.

Reply:
ack